### DISCUSSION TITLE
What am I doing wrong that the size of my data account is 0 when passed to program via keys?

### DISCUSSION THREAD

**Discussion by nukec** [bmresearch/Solnet#191] [Q&A]
I have simple rust program with following code:
```
let game_account = next_account_info(account_info_iter)?;
let dorito = &game_account.try_data_len()?;

msg!("size of dorito: {:?}", dorito);
```

The code above gives me the size of dorito - data account (game_account) of 0, while when if I use solana-web3.js I have no issues whatsoever, I get the full size of data account, aka account is passed properly.

But with solnet with following code the size of data account is 0, which means I am probably missing something:
```
public async Task<PublicKey?> SetNewGameDataAsync(PublicKey playerPublicKey, PublicKey dataAccountPublicKey)
        {
            RequestResult<ResponseValue<BlockHash>>? blockHash = await this._rpcClient.GetRecentBlockHashAsync();

            PublicKey programId = new(this._chainConfiguration.ProgramId);

            List<AccountMeta> keys = new()
                                     {
                                         AccountMeta.Writable(publicKey: dataAccountPublicKey, isSigner: false),
                                         AccountMeta.ReadOnly(publicKey: playerPublicKey, isSigner: false),
                                         AccountMeta.ReadOnly(new PublicKey(this._chainConfiguration.ServerAddressPublicKey), isSigner: false)
                                     };

            byte[]? tx = new TransactionBuilder().SetRecentBlockHash(blockHash.Result.Value.Blockhash)
                                                 .SetFeePayer(this._account)
                                                 .AddInstruction(new TransactionInstruction { ProgramId = programId, Keys = keys, Data = GameCommands.CreateGame() })
                                                 .Build(this._account);

            RequestResult<ResponseValue<SimulationLogs>>? simulatedTransaction = await this._rpcClient.SimulateTransactionAsync(tx);

            if (!simulatedTransaction.WasSuccessful)
            {
                this._logger.LogError("Solana cluster " + this._chainConfiguration.Cluster + ": Could not simulate transaction for creating new game - Server code " +
                                      simulatedTransaction.ServerErrorCode);

                return null;
            }

            await this._rpcClient.SendTransactionAsync(tx);

            return dataAccountPublicKey;
        }
```

**Comment by hoakbuilds** [comment]
Difficult to know for sure without more context into the rust code.
 You are not passing any data in the instruction, is this intended?

I am not really seeing anything in the code that is standing out as being wrong.

**Reply by nukec** [reply]
I am passing `Data = GameCommands.CreateGame()` which is just instruction command that I was testing, but doesn't relate to this issue:
```        public static byte[] CreateGame()
        {
            byte[] commandBuffer = new byte[8];
            commandBuffer.WriteU8((byte)GameCommand.CREATE_GAME, offset: 0);

            return commandBuffer;
        }
```

Rust program is this literally very simple trying to check the size of data account:
```
pub fn process_instruction(
    program_id: &Pubkey,
    accounts: &[AccountInfo],
    instruction_data: &[u8],
) -> ProgramResult {
    msg!("Helloworld Rust program entrypoint");

    
    let account_info_iter = &mut accounts.iter();

        let game_account = next_account_info(account_info_iter)?;
        let player_account = next_account_info(account_info_iter)?;
        let server_account = next_account_info(account_info_iter)?;

        msg!("game account key: {:?}", game_account.key);
        msg!("player x account key: {:?}", player_account.key);
        msg!("server account key: {:?}", server_account.key);

        let dorito = &game_account.try_data_len()?;

        msg!("size of dorito: {:?}", dorito);
Ok(());
```

> {
>   "jsonrpc": "2.0",
>   "result": {
>     "context": {
>       "slot": 75982330
>     },
>     "value": {
>       "accounts": null,
>       "err": null,
>       "logs": [
>         "Program HmaajdGsXrRPdCqkA5Cv4Nm5eKrUrUK5MFwu14WuD3Gq invoke [1]",
>         "Program log: Helloworld Rust program entrypoint",
>         "Program log: game account key: 33AFjgF6Cvw9wZvre7uWyoEok9KxSQf8NctRbDCtpqDP",
>         "Program log: player x account key: 5Z9oJEs2hUBd5vNTDWZq4A7NANAkPeBRRjQw7eEvbPwB",
>         "Program log: server account key: FbSwjbBBKPzUiPzK9SewRnVTAK1USEXf9xvv248c2zAm",
>         "Program log: size of dorito: 0",
>         "Program log: done",
>         "Program HmaajdGsXrRPdCqkA5Cv4Nm5eKrUrUK5MFwu14WuD3Gq consumed 36910 of 200000 compute units",
>         "Program HmaajdGsXrRPdCqkA5Cv4Nm5eKrUrUK5MFwu14WuD3Gq success"
>       ]
>     }
>   },
>   "id": 7
> }

Following code is in typescript:
```
export async function createNewGame(): Promise<void> {
  console.log(
    "Creating a game on ",
    gameAccountPubKey.toBase58(),
    " for player ",
    playerXAccount.publicKey.toBase58()
  );
  serverAccountPubKey = new PublicKey(
    "FbSwjbBBKPzUiPzK9SewRnVTAK1USEXf9xvv248c2zAm"
  );

  console.log(programId);
  const instruction = new TransactionInstruction({
    keys: [
      { pubkey: gameAccountPubKey, isSigner: false, isWritable: true },
      { pubkey: playerXAccountPubKey, isSigner: false, isWritable: true },
      { pubkey: serverAccountPubKey, isSigner: false, isWritable: true },
    ],
    programId,
    data: createGameCommand(),
  });
  await sendAndConfirmTransaction(
    connection,
    new Transaction().add(instruction),
    [payerAccount],
    {
      commitment: "singleGossip",
      preflightCommitment: "singleGossip",
    }
  );
}
```

And I am getting the size no problem:

>  Status: Ok
>   Log Messages:
>     Program HmaajdGsXrRPdCqkA5Cv4Nm5eKrUrUK5MFwu14WuD3Gq invoke [1]
>     Program log: Helloworld Rust program entrypoint
>     Program log: game account key: 33AFjgF6Cvw9wZvre7uWyoEok9KxSQf8NctRbDCtpqDP
>     Program log: player x account key: 5Z9oJEs2hUBd5vNTDWZq4A7NANAkPeBRRjQw7eEvbPwB
>     Program log: server account key: FbSwjbBBKPzUiPzK9SewRnVTAK1USEXf9xvv248c2zAm
>     Program log: size of dorito: 307
>     Program log: size of dorito 2: 307
>     Program log: done
>     Program HmaajdGsXrRPdCqkA5Cv4Nm5eKrUrUK5MFwu14WuD3Gq consumed 37706 of 200000 compute units
>     Program HmaajdGsXrRPdCqkA5Cv4Nm5eKrUrUK5MFwu14WuD3Gq success

PS: I generate new account each time for same program Id, so it's same code, i am not coind

**Reply by hoakbuilds** [reply]
I am really unsure what might be wrong here, I have never encountered this issue when interacting with way more complex programs. The only thing I see different between the Solnet and the web3.js snippets is the fact that web3.js is passing all accounts as `Writable` while the solnet code does not.

**Reply by nukec** [reply]
if you deploy very simple program like mine and you try to pass data account via keys via solnet code, does it work for you? that's quite crucial functionality which is not covered in examples and tests i think. so creating data account works without issues with solnet, but when i am trying to pass created data account to program via Keys then size is 0. would it be possible for you to try this with very simple program like this(rust program):
```
pub fn process_instruction(
    program_id: &Pubkey,
    accounts: &[AccountInfo],
    instruction_data: &[u8],
) -> ProgramResult {
    msg!("Helloworld Rust program entrypoint");

    
    let account_info_iter = &mut accounts.iter();

        let data_account = next_account_info(account_info_iter)?;

        msg!("data account key: {:?}", data_account.key);

        let sizeOfDataAccount = &data_account.try_data_len()?;

        msg!("size of data account: {:?}", sizeOfDataAccount);
Ok(());
```

**Reply by hoakbuilds** [reply]
I see another big difference in the code snippets. In web3.js you are sending the transaction and waiting for confirmation while in the Solnet code you are only simulating the transaction, have you tried submitting the transaction normally and seeing if this behavior persists?

I am going to attempt replicating this.

**Reply by hoakbuilds** [reply]
Using the same program you deployed and account keys you have shown here for the example I am not being able to replicate this issue.

https://solscan.io/tx/5JzSHpKnXDcZHbvZ5Gt6okJcUhNZuYMPtrBtjFB3NfPFiBUqjLbd2MdzuAmJJmhB88xmws7bkeuBAGyU6ZjptyLw?cluster=devnet

**Reply by hoakbuilds** [reply]
Are you creating an account and attempting to execute this instruction sequentially? It could be that you are trying to do this in a really short timespan and the transaction which creates the account which you want to get the account data of hasn't been finalized.

**Reply by hoakbuilds** [reply]
This last comment may be incorrect, I have managed to create an account and executing the instruction in the same transaction, as you can see here.

https://solscan.io/tx/cBVz4JqHqKmYCsV6EpbwWA7wEY6Shunb2AK7uhkyJpMaQTCvVqA9rzVqV36oA4nAWT3hVumzwwCzAg44oRzSBYV

The code I have used to attempt replicate this:
https://gist.github.com/murlokito/319bd6e941f418d188f8f66fbbe10012
